=htc Refactor PoolSlot to GraphStage to address race conditions #19643

leachbj opened the pull request:
The previous implementation had a number of race conditions which
could result in exceptions or a hung client.

leachbj committed:
=htc Refactor PoolSlot to GraphStage to address race conditions #19643

The previous implementation had a number of race conditions which
could result in exceptions or a hung client.

leachbj commented:
this is still WIP.  Pretty sure this version is "working" but there are some parts that don't really seem like the right way.

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -89,187 +76,163 @@ private object PoolSlot {
+          val delivery = ResponseDelivery(ResponseContext(requestContext, Success(response withEntity entity)))
+          import fm.executionContext
+          val requestCompleted = SlotEvent.RequestCompletedFuture(whenCompleted.map(_ ⇒ SlotEvent.RequestCompleted(slotIx)))
+          push(out, delivery :: requestCompleted :: Nil)
```
Do this instead:

```scala
emitMultiple(out, delivery :: requestCompleted :: Nil)
```

Then we can make the output type not a `List[]` anymore, which allows us to remove:

```scala
.mapConcat(ConstantFun.scalaIdentityFunction))
```

So win-win

drewhk commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,37 +38,25 @@ private object PoolSlot {
-                                                                                                 v
+    Request-   +-----------+              +-------------+              +------------+
+    Context    | Slot-     |  List[       |   flatten   |  Processor-  | SlotEvent- |  Response-
+    +--------->| Processor +------------->| (MapConcat) +------------->| Split      +------------->
```
Does this whole pipeline make any sense still? I mean the SlotProcessor+flatten chain? Can't be these 3 stages collapsed into one?

ktoso replied:
That's where I was heading btw - we can get rid of the concat easily too, see hints below

drewhk replied:
No need to do it all at once though, go the least risky way. We can continue the cleanup after a bugfix release.

ktoso replied:
I agree 👍

drewhk commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -89,187 +76,163 @@ private object PoolSlot {
-     */
-    override def receive = waitingExposedPublisher
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler { self ⇒
+      private var inflightRequests = immutable.Queue.empty[RequestContext]
```
Please never use immutable.Queue.empty. It is slow and immutability does not gain anything for us here. Please use a java.util.ArrayDeque

ktoso replied:
I think I was benchmarking this actually here and saw no difference.
But you're right, let's try the Deque in any case and see (Queue was in previous impl already)

That's a low priority part of the PR

drewhk replied:
I did benchmarking elsewhere and it does make quite a bit of difference, especially allocations. Also, this will be now more fused than before, so maybe we can have more speed.

ktoso replied:
Now that you mention it, my bench was probably not on a fused one.
I agree that Deque should theoretically be better, let's give it a shot while we're here

drewhk commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -89,187 +76,163 @@ private object PoolSlot {
-        connOutport ! Request(totalDemand)
-        context.become(waitingForDemandFromConnection(connInport = connInport, connOutport = connOutport, rc))
+      // request first request/command
+      override def preStart(): Unit = pull(in)
```
This makes the Flow detached. Are we sure we don't want to make it non-detached if possible? (i.e. only pull in response to downstream pull)

drewhk commented:
OK TO TEST

drewhk commented:
I would like to see the full duty-cycle of this new stage. Currently it looks like a detached-one (i.e. two, mostly concurrent cycles) but I am not sure that is needed. I would imagine something like this:
 - downstream pulls we pull inner stream
 - inner stream pulls we pull upstream
 - upstream pushes, we push inner stream
 - inner stream pushes we push downstream

Currently the stage immediately pulls and that can complicate matters. While sometimes this is necessary I am not convinced this is necessary for this stage.

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,37 +38,25 @@ private object PoolSlot {
+                                                                             |  via slotEventMerge)
+                                                                             v
    */
   def apply(slotIx: Int, connectionFlow: Flow[HttpRequest, HttpResponse, Any])(implicit system: ActorSystem, fm: Materializer): Graph[FanOutShape2[SlotCommand, ResponseContext, RawSlotEvent], Any] =
```
As first step this can be simplified as

```scala
def apply(slotIx: Int, connectionFlow: Flow[HttpRequest, HttpResponse, Any])(implicit m: Materializer) = 
    GraphDSL.create() { implicit b ⇒
      import GraphDSL.Implicits._

      val log = ActorMaterializerHelper.downcast(m).logger
      val slotProcessor: FlowShape[SlotCommand, ProcessorOut] = b.add(new SlotProcessor(slotIx, connectionFlow, log))
      val partition = b.add(Partition[ProcessorOut](2, {
        case ResponseDelivery(r) ⇒ 0
        case r: RawSlotEvent ⇒ 1
      }))
      slotProcessor.out ~> partition.in

      val deliveries = partition.out(0).map(_.asInstanceOf[ResponseDelivery])
      val events = partition.out(1).map(_.asInstanceOf[RawSlotEvent])
      new FanOutShape2[SlotCommand, ResponseDelivery, RawSlotEvent](slotProcessor.in, deliveries.outlet, events.outlet)
    }
```

ktoso replied:
2nd step is to actually have it all in 1 stage, no reason to do so much dancing here, slot processor could push into right places right away

leachbj replied:
Is that really simplified?  The split at least avoids the `asInstanceOf`.

ktoso replied:
Yeah, perhaps not really better. Step 2 would be, if we make it all one stage

leachbj committed:
Remove extraneous stages previously supporting PoolSlot

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -38,67 +36,41 @@ private object PoolSlot {
-    }
+  def apply(slotIx: Int, connectionFlow: Flow[HttpRequest, HttpResponse, Any])(implicit system: ActorSystem /*REMOVEME*/ , m: Materializer): Graph[FanOutShape2[SlotCommand, ResponseContext, RawSlotEvent], Any] = {
+    val log = ActorMaterializerHelper.downcast(m).logger
+    new SlotProcessor(slotIx, connectionFlow, log)
```
exactly, great :)

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -38,67 +36,41 @@ private object PoolSlot {
-        split.out(0).collect { case ResponseDelivery(r) ⇒ r }.outlet,
-        split.out(1).collect { case r: RawSlotEvent ⇒ r }.outlet)
-    }
+  def apply(slotIx: Int, connectionFlow: Flow[HttpRequest, HttpResponse, Any])(implicit system: ActorSystem /*REMOVEME*/ , m: Materializer): Graph[FanOutShape2[SlotCommand, ResponseContext, RawSlotEvent], Any] = {
```
remove system right away then if aware that `/*REMOVEME*/`? :)

leachbj replied:
i've been cheating with my testing and needed the binary compatibility so was leaving it in.

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -38,67 +36,41 @@ private object PoolSlot {
    */
   private class SlotProcessor(slotIx: Int, connectionFlow: Flow[HttpRequest, HttpResponse, Any], log: LoggingAdapter)(implicit fm: Materializer)
-    extends GraphStage[FlowShape[SlotCommand, List[ProcessorOut]]] {
+    extends GraphStage[FanOutShape2[SlotCommand, ResponseContext, RawSlotEvent]] {
```
👍

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -38,67 +36,41 @@ private object PoolSlot {
 
     val in: Inlet[SlotCommand] = Inlet("SlotProcessor.in")
-    val out: Outlet[List[ProcessorOut]] = Outlet("SlotProcessor.out")
+    val out0: Outlet[ResponseContext] = Outlet("SlotProcessor.response-out")
```
`responseOut` or `responsesOut`?

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -38,67 +36,41 @@ private object PoolSlot {
     val in: Inlet[SlotCommand] = Inlet("SlotProcessor.in")
-    val out: Outlet[List[ProcessorOut]] = Outlet("SlotProcessor.out")
+    val out0: Outlet[ResponseContext] = Outlet("SlotProcessor.response-out")
+    val out1: Outlet[RawSlotEvent] = Outlet("SlotProcessor.event-out")
```
`eventsOut`?

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -224,14 +201,21 @@ private object PoolSlot {
+        override def onPull(): Unit = log.debug("BERN-{}: PoolSlot: onPull(out0)", slotIx)
+      })
+      setHandler(out1, new OutHandler {
+        @scala.throws[Exception](classOf[Exception])
```
remove `@throws`

ktoso commented:
Status here FYI team: This is a good direction but does not handle all things yet. Likely good to merge as base and put fixes on top of it, sadly I did not yet manage to make those fixes a reality :<

leachbj committed:
Make the flow connected, tie in SlotCommand to RawSlotEvents out

leachbj committed:
Limit elements to downstream to 1 outstanding

Since connectionFlow uses One2OneBidiFlow its strictly one in, one out
and since any request can close the connection we can't stack them as
that would result in a OutputTruncationException.

Upstream demand is now only generated when a response is completed or
the substream completes.

leachbj commented:
With the addition of d7d79bb the new tests in https://github.com/akka/akka/pull/21321 are passing but that is because we're basically disabling pipeling within the slot.  The issue is the pipelining code will happy send down GET / & POST /foo to the slot, the GET can close the connection but the POST will already be in the connectionFlow and cause it to fail with `OutputTruncationException` which in this example would result in the failure hitting the caller since the POST cannot be retried (it would happen in GET / & GET / case too if the 2nd one had run out of retries).

There are two failures;
should never close hot connections when minConnections key is given and >0 (minConnections = 1)
should never close hot connections when minConnections key is given and >0 (minConnections = 5) which both timeout waiting for `TestSubscriber$OnNext`.

leachbj committed:
Additional connection closed tests which reproduce #19643

leachbj commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +37,154 @@ private object PoolSlot {
+            val it = inflightRequests.descendingIterator()
+            while (it.hasNext) {
+              val rc = it.next()
+              if (rc.retriesLeft == 0) { println(s"failing $rc"); emit(out0, ResponseContext(rc, Failure(ex))) }
```
remove debug

leachbj committed:
Add missing pull on eager connection setup

2beaucoup commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -335,6 +335,70 @@ class ConnectionPoolSpec extends AkkaSpec("""
+  "xoxo be able to handle 500 pipelined requests with connection termination" in new TestSetup(autoAccept = true) {
+    def closeHeader(): List[Connection] =
+      //      if (util.Random.nextInt(8) == 0)
+      Connection("close") :: Nil
```
Shouldn't we test some non-closing requests/responses in between, too?

leachbj replied:
Yes - adding that showed up more problems!

leachbj commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -335,6 +335,70 @@ class ConnectionPoolSpec extends AkkaSpec("""
+    Await.result(idSum, 10.seconds) shouldEqual N * (N + 1) / 2
+  }
+
+  "xoxo be able to handle 500 pipelined requests with connection termination" in new TestSetup(autoAccept = true) {
```
xoxo - anything else we need to do to finalise this test?

ktoso replied:
you can remove the `xoxo` btw ;) That's only a trick I use to `testOnly *Clazz* -- -z xoxo` to run a specific test

2beaucoup commented on akka-http-core/src/test/scala/akka/http/impl/engine/client/ConnectionPoolSpec.scala:
```diff
@@ -335,6 +335,70 @@ class ConnectionPoolSpec extends AkkaSpec("""
+
+      try {
+        val N = 200
+        info(s"n=${N}, poolFlow=${poolFlow}")
```
remove?

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +37,156 @@ private object PoolSlot {
+            Source.single(rc.request).concat(Source.fromGraph(connectionFlowSource.source))
+              .via(connectionFlow).runWith(Sink.fromGraph(connectionFlowSink.sink))(subFusingMaterializer)
+
+            connectionFlowSink.pull()
```
Should we stop accepting new requests if this one has `Connection: close`?

leachbj replied:
Probably could, but thats only half the story since the server might decide to close the connection anyhow.

2beaucoup replied:
But why dispatch more requests onto a known-to-get-closed connection although we know they'll take a retry cycle anyway?

leachbj committed:
Resolve test failure when using a mix of "Connection: Close" requests

leachbj committed:
Fixed failedRequests value in Disconnect response

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +35,133 @@ private object PoolSlot {
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          val retryRequests = inflightRequests.iterator().asScala.filter(_.retriesLeft > 0)
+            .map(rc ⇒ SlotEvent.RetryRequest(rc.copy(retriesLeft = rc.retriesLeft - 1))).toList
+          val failures = inflightRequests.iterator().asScala.filter(_.retriesLeft == 0).map(rc ⇒ ResponseContext(rc, Failure(ex))).toList
```
```scala
val (retries, failures) = inflightRequests.iterator().asScala.partition(_.retriesLeft > 0)
```

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +35,133 @@ private object PoolSlot {
-      case FromConnection(OnError(e)) ⇒ handleDisconnect(sender(), Some(e))
-    }
+      override def onPush(): Unit = {
+        def createSubSourceOutlets() = {
```
`createSubFlowPorts` or something alike?

2beaucoup commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +35,133 @@ private object PoolSlot {
+          case DispatchCommand(rc: RequestContext) ⇒
+            if (isConnected) {
+              inflightRequests.add(rc)
+              connectionFlowSource.push(rc.request)
```
Again: I think we should signal that we don't accept further requests `if (rc.request.header[Connection].exists(_.hasClose))`.

jrudolph commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +35,133 @@ private object PoolSlot {
-      .toMat(Sink.actorSubscriber[HttpResponse](flowOutportProps(self)))(Keep.both)
-      .named("SlotProcessorInternalConnectionFlow")
+    val in: Inlet[SlotCommand] = Inlet("SlotProcessor.in")
+    val out0: Outlet[ResponseContext] = Outlet("SlotProcessor.responsesOut")
```
Could be renamed to "responsesOut".

jrudolph commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +35,133 @@ private object PoolSlot {
-      .named("SlotProcessorInternalConnectionFlow")
+    val in: Inlet[SlotCommand] = Inlet("SlotProcessor.in")
+    val out0: Outlet[ResponseContext] = Outlet("SlotProcessor.responsesOut")
+    val out1: Outlet[RawSlotEvent] = Outlet("SlotProcessor.eventsOut")
```
Could be renamed to "eventsOut"

leachbj committed:
Align the inner and out flow demands

leachbj commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +35,135 @@ private object PoolSlot {
-      case ev @ (OnComplete | OnError(_)) ⇒
-        slotProcessor ! FromConnection(ev)
-        context.stop(self)
+          connectionFlowSink.pull()
```
I don't think we want to pull here, I think that should come when there is demand at the connectionFlowSource but I wasn't able to get that to work.

leachbj commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +35,135 @@ private object PoolSlot {
-        onNext(SlotEvent.ConnectedEagerly(slotIx) :: Nil)
-        out ! Request(totalDemand)
-        context.become(waitingEagerlyConnected(connInport = in, connOutport = out))
+          val (inflightRetry, inflightFail) = if (ex.isEmpty) (inflightRequests.asScala, Nil)
```
This says if there wasn't a failure on the inner stream then the inflightRequest was never sent so just bounce it back.  I don't think we ever should have pulled this request in the first place but I haven't been able to get the demand request right.

jrudolph replied:
That's exactly the pipelining support of the lower layers in `OutgoingConnectionBlueprint` which will pull new requests even while another one is still in flight and send these out as well. So, in fact, a new request may or may not been have sent out. As soon as you push it to the connection it will try to send it out. This may happen concurrently with the server sending back the previous response and closing the connection.

jrudolph commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +35,135 @@ private object PoolSlot {
+          else inflightRequests.iterator().asScala.partition(_.retriesLeft > 0)
 
-      case Request(_) ⇒ if (remainingRequested == 0) request(1) // ask for first request if necessary
+          val retries = inflightRetry.map(rc ⇒ SlotEvent.RetryRequest(if (rc.retriesLeft > 0) rc.copy(retriesLeft = rc.retriesLeft - 1) else rc)).toList
```
Wouldn't this retry requests regardless of the number of retries left if we hit the `ex.isEmpty` branch above?

leachbj replied:
Yeah, was rushing too much this should just say if ex is empty don't decrement because the request was never actually tried.

leachbj replied:
Sorry, yes if ex.isEmpty then we should always retry the request regardless of the number of retries since the request was never actually issued.  The substream will request one, then when that is sent and a response is read it will request a second.  If that first request did a connection close then the stream just completes without an error, theres no network traffic out.

The retry logic applies if there is a connection refused (although that case is a bit silly since the retries are super fast its unlikely that there'd be any change) or if there's a network error during the request write or the response header read (a response body read will show up on the entity stream as an error unless we have a `Strict` response).

I've got a small refactor here that makes this a bit clearer and remvoes the `UnexpectedDisconnectException` since that actually can't happen.

jrudolph replied:
> Sorry, yes if ex.isEmpty then we should always retry the request regardless of the number of retries since the request was never actually issued.

Is that really the case? How did the request then get into the inflight requests? I'll have another look...

jrudolph replied:
> The substream will request one, then when that is sent and a response is read it will request a second.

Yes, that's the flow with persistent HTTP connections. However, if pipelining would work, requests would be dispatched without waiting for the response of a previous request. So, connection closing could also run concurrently with ongoing requests.

jrudolph commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala:
```diff
@@ -36,240 +35,135 @@ private object PoolSlot {
+      setHandler(responsesOut, new OutHandler {
+        override def onPull(): Unit = {
+          // downstream pulls, if connected we pull inner
+          if (isConnected) connectionFlowSink.pull()
```
Isn't this here prone to double pulling as well?

leachbj replied:
the `isConnected` case doesn't have this problem since we have a simple 1-1, the guard is on the disconnected because the `eventsOut` doesn't always get an element.

jrudolph replied:
I see, thanks for the explanation.

jrudolph commented:
Thanks again, @leachbj. I think this PR is a great step forward in that it improves the code of PoolSlot a lot. It's perfect that the changes are contained in this one file and that we add more tests.

That said, I am not particularly confident that the existing tests are covering everything (e.g. things I pointed out in the comments are likely bugs that are not yet caught). In particular, we should make sure:
 * that pipelining itself (still) works, i.e. that multiple requests are scheduled onto one connection
 * that all those pipelining edge cases are handled correctly, i.e. an idempotent / non-idempotent request is scheduled to one connection concurrently with connection closure / failure and retries work when necessary
 * no double or missed pulls

I will review the existing tests now and add ones I find we should have. If everything is green we should rather merge sooner than later and then follow up with further cleanup.

leachbj committed:
Simplify PoolSlot retry to eliminate UnexpectedDisconnectException

This exception wasn't being thrown and was only a placeholder
for a path that wouldn't happen.

ktoso commented:
Thanks for all the work on this one @leachbj.
I think that the remaining (pipelining mostly) things can be addressed separately, and for now let's merge this and do a release as we have some other bugfixes we want to ship as well together with it.

LGTM

jrudolph commented:
I think the implementation currently relies on support for low-level client-side HTTP pipelining being broken right now (see [this comment](https://github.com/akka/akka/pull/21316#discussion_r77504676)). 

Also, with this PR (and probably also before), the `PoolConductor` will schedule requests onto a slot as if it was able to pipeline and the request will then be buffered in the PoolSlot (as the low-level implementation will not fetch more than one request at a time). AFAICS the effect is mostly that, if a user tries to enable pipelining, we will fetch and buffer more requests than necessary in PoolSlots that would otherwise be backpressured at the request source. This may result in less optimal scheduling decisions (buffered requests may be stuck in a slow slot that would otherwise have had the chance to be later scheduled on another slot). But this is not a correctness issue.

So, I think we should merge this to move this forward and then follow up with a ticket that fixes (or removes) pipelining and adds more tests that everything is working as intended.

LGTM

jrudolph commented:
Could you squash your commits into one and also reference all the issues it probably fixes: #19643, #20570, #21235?

ktoso commented:
LGTM, great work - thanks @leachbj !

ktoso commented:
PS: We can squash and reword things using github UI nowadays :)
